common/djangoapps/util/password_policy_validators.py

```diff
@@ -1,322 +1,384 @@
+            return ''
 
-    _validate_password_dictionary(password)
-    _validate_password_against_username(password, username)
 
-    # Some messages are composable, so we'll add them together here
-    errors = [_validate_password_length(password)]
-    errors += _validate_password_complexity(password)
-    errors = filter(None, errors)
+class NumericValidator(object):
+    """
+    Validate whether the password contains at least min_numeric numbers.
 
-    if errors:
-        msg = _('Enter a password with at least {requirements}.').format(requirements=' & '.join(errors))
-        raise ValidationError(msg)
+    Parameters:
+        min_numeric (int): the minimum number of numeric characters to require
+            in the password. Must be >= 0.
+    """
+    def __init__(self, min_numeric=0):
+        self.min_numeric = min_numeric
+
+    def validate(self, password, user=None):
+        count = 0
+        for character in password:
+            if count == self.min_numeric:
+                return
+            if character.isnumeric():
+                count += 1
+        raise ValidationError(
+            ungettext(
+                'Your password must contain at least %(min_numeric)d number.',
+                'Your password must contain at least %(min_numeric)d numbers.',
+                self.min_numeric
+            ),
+            code='too_few_numeric_char',
+            params={'min_numeric': self.min_numeric},
+        )
+
+    def get_help_text(self):
+        return ungettext(
+            "Your password must contain at least %(min_numeric)d number.",
+            "Your password must contain at least %(min_numeric)d numbers.",
+            self.min_numeric
+        ) % {'min_numeric': self.min_numeric}
+
+    def get_instruction_text(self):
+        if self.min_numeric > 0:
+            return ungettext(
+                '%(num)d number',
+                '%(num)d numbers',
+                self.min_numeric
+            ) % {'num': self.min_numeric}
+        else:
+            return ''
 
 
-def _validate_password_security(password, user):
+class UppercaseValidator(object):
     """
-    Check password reuse and similar operational security policy considerations.
+    Validate whether the password contains at least min_upper uppercase letters.
+
+    Parameters:
+        min_upper (int): the minimum number of uppercase characters to require
+            in the password. Must be >= 0.
     """
-    # Check reuse
-    if not PasswordHistory.is_allowable_password_reuse(user, password):
-        if user.is_staff:
-            num_distinct = settings.ADVANCED_SECURITY_CONFIG['MIN_DIFFERENT_STAFF_PASSWORDS_BEFORE_REUSE']
+    def __init__(self, min_upper=0):
+        self.min_upper = min_upper
+
+    def validate(self, password, user=None):
+        count = 0
+        for character in password:
+            if count == self.min_upper:
+                return
+            if character.isupper():
+                count += 1
```

Comment:
This same block of code appears several times. It might be a good candidate for a utility function. And maybe the block could be written more briefly/readably like the following. A matter of taste, but the iter loop adding to count felt un-pythonic to me, personally.

```
valid_count = len([c for c in password if c.isupper()])
if valid_count >= self.min_upper:
  return
```